omnimatter_compression/function.lua

```diff
@@ -60,61 +60,48 @@ end
 -----------------------------------------------------------------------------
 -- ICON FUNCTIONS --
 -----------------------------------------------------------------------------
---Find base icon
+--If our prototype doesn't have an icon, we need to find one that does
 local find_icon = function(item)
   for _, p in pairs({"item","mining-tool","gun","ammo","armor","repair-tool","capsule","module","tool","rail-planner","item-with-entity-data","fluid"}) do
     if data.raw[p][item] then
-      if data.raw[p][item].icons then
-        return data.raw[p][item].icons
-      else
-        return {{icon = data.raw[p][item].icon, icon_size = data.raw[p][item].icon_size or 32}}
+      if data.raw[p][item].icons or data.raw[p][item].icon then
+        return data.raw[p][item]
       end
     end
   end
 end
 --really dig deep for the icon set
-local find_result_icon = function(it)
-  if type(it) == "table" then
-    if it then --search for the icon in its various hiding places
-      if it.icons then
-        icons=table.deepcopy(it.icons)
-        if it.icon_size and not it.icons[1].icon_size then
-          icons[1].icon_size=it.icon_size or 32
-        end
-        return icons
-      else
-        if it.icon then
-          return  {{icon = it.icon, icon_size = it.icon_size or 32}}
-        else
-          local process = {}
-          if it.result or (it.normal and it.normal.result) then
-            if it.result then
-              process = find_icon(it.result)
-            else
-              process = find_icon(it.normal.result)
-            end
-          else
-            if it.results then
-              process = find_icon(it.results[1].name)
-            else
-              process = find_icon(it.normal.results[1].name)
-            end
-          end
-          if process ~= nil then
-            if process[1] and process[1].icon then
-              if it.icon_size and not process[1].icon_size then
-                process[1].icon_size=it.icon_size or 32
-              end
-            end
-            return process
-          else
-            return {{icon = "__omnimatter_compression__/graphics/compress-out-arrow-32.png", icon_size = 32}} --should be using blank, but for testing...
-          end
-        end
+local function find_result_icon(raw_item)
+  if raw_item then
+    if type(raw_item) ~= "table" then
+      raw_item = find_icon(raw_item) --Find a matching prototype if possible
+      return find_result_icon(raw_item)
+    elseif raw_item.icons then
+      local icons = table.deepcopy(raw_item.icons)
+      for i, icon in pairs(icons) do-- Apply inherited attributes as explicit for each layer
+        icon.icon_size = icon.icon_size or raw_item.icon_size
+        icon.icon_mipmaps = icon.icon_mipmaps or raw_item.icon_mipmaps
       end
+      return icons
+    elseif raw_item.icon then
+      return {{
+        icon = raw_item.icon,
+        icon_size = raw_item.icon_size
```

Comment:
Shouldn't this also check for icon_mipmaps?

Reply:
According to IconSpecification if it's _icon_ not _icons_, it doesn't support the mipmap attribute.

Reply:
I don't think that's true. That's possibly an error with the wiki, a lot of vanilla icons (all of them?) have icon_mipmaps specified and they use almost exclusively icon.

Reply:
Fixed in c640e00